mlir/lib/Bindings/Python/IRAttributes.cpp

```diff
@@ -997,6 +1022,56 @@ class PyDenseIntElementsAttribute
   }
 };
 
+class PyDenseResourceElementsAttribute
+    : public PyConcreteAttribute<PyDenseResourceElementsAttribute> {
+public:
+  static constexpr IsAFunctionTy isaFunction =
+      mlirAttributeIsADenseResourceElements;
+  static constexpr const char *pyClassName = "DenseResourceElementsAttr";
+  using PyConcreteAttribute::PyConcreteAttribute;
+
+  static PyDenseResourceElementsAttribute
+  getFromBuffer(py::buffer buffer, std::string name, PyType type,
+                DefaultingPyMlirContext contextWrapper) {
+    // Do not request any conversions as we must ensure to use caller
+    // managed memory.
+    int flags = 0;
+    Py_buffer view;
+    if (PyObject_GetBuffer(buffer.ptr(), &view, flags) != 0) {
+      throw py::error_already_set();
+    }
+    auto freeBuffer = llvm::make_scope_exit([&]() { PyBuffer_Release(&view); });
```

Comment:
`AsmResourceBlob` has a deleter callback, could that be used to delay this call and relax the "caller arranges to keep the backing memory alive" restriction?

Reply:
You know, I was so focused on the CAPI side, which doesn't expose that hook, that I completely overlooked that. Too focused on one side of the API.

I could just make a new entry point like this one (but exposing the deleter callback):

```
MLIR_CAPI_EXPORTED MlirAttribute mlirUnmanagedDenseBlobResourceElementsAttrGet(
    MlirType shapedType, MlirStringRef name, const void *data,
    size_t dataLength) {
  return wrap(DenseResourceElementsAttr::get(
      llvm::cast<ShapedType>(unwrap(shapedType)), unwrap(name),
      UnmanagedAsmResourceBlob::allocateInferAlign(
          llvm::ArrayRef(static_cast<const char *>(data), dataLength))));
}
```
Even better, I could just change that one. I just landed it a day or two ago and was cargo-culting the others. I think we can still change the API.

Reply:
I could have sworn we had resource support C API side much earlier than that.

Reply:
How about that? I don't know how to further test it from the Python API, but I did add a printf and made sure it was releasing.

Reply:
> I could have sworn we had resource support C API side much earlier than that.

There is some but it was not general/complete. I didn't find any of the existing CAPI for it to be workable for this situation.

Reply:
I figured out how to test the memory freeing from Python.

Reply:
(also fixed a datatype in this file that was causing numpy to log warnings about deprecated conversions)